### PR TITLE
merge v4 into redwood

### PR DESCRIPTION
**Pull request by wgwz**
## Description

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] provided a link to the relevant issue or specification
- [ ] provided instructions on how to test
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed
- [ ] once the PR is closed, set up backport PRs for `redwood` and `hambach` (see below)

#### Setting up backport PRs

After merging your PR to `master`, set up backports by doing the following:

1. If your branch does not have merge commits, add the following comment to
   your PR, `@Mergifyio backport redwood hambach`.

2. If your branch does have merge commits:

    a. Pull latest `master`, `hambach` and `redwood` branches

    b. Create new branches for backports and merge `master` (replace `<PR#>` with your PR #)
    ```
    git checkout -b hambach-backport-<PR#> hambach
    git merge master
    git push origin hambach-backport-<PR#>
    git checkout -b redwood-backport-<PR#> redwood
    git merge master
    git push origin redwood-backport-<PR#>`
    ```

    c. Open new PRs in regen-web targeting `hambach` and `redwood`, respectively.
  

### How to test

1.

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**._

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] verified React components follow DRY principles
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

### CODE REVIEW

File: web-registry/src/lib/ecocredit/api.ts [hunk 1]
```diff
@@ -530,10 +598,136 @@ export const queryCreditTypes = async ({
...
+    }
+  } catch (err) {
+    throw new Error(`Error fetching batches: ${err}`);
+  }
+};
+
+export const isVCSCreditClass = (creditClassId?: string): boolean => {
+  // TODO: this is a hack to make V4 testnet work. C02 is a copy of C01 on V4.
+  // In PROD/mainnet, this should only be C01.
+  return creditClassId === 'C01' || creditClassId === 'C02';
```
Comment (mhagel) [inline web-registry/src/lib/ecocredit/api.ts:732]:
```suggestion
  return creditClassId === 'C01';
```

File: web-registry/src/lib/ecocredit/api.ts [hunk 2]
```diff
@@ -530,10 +598,136 @@ export const queryCreditTypes = async ({
...
+      return batchInfos;
+    }
+  } catch (err) {
+    throw new Error(`Error fetching batches: ${err}`);
+  }
+};
+
+export const isVCSCreditClass = (creditClassId?: string): boolean => {
+  // TODO: this is a hack to make V4 testnet work. C02 is a copy of C01 on V4.
+  // In PROD/mainnet, this should only be C01.
```
Comment (mhagel) [inline web-registry/src/lib/ecocredit/api.ts:731]:
```suggestion
```

File: web-registry/src/lib/ledger.ts [hunk 3]
```diff
@@ -6,8 +6,15 @@ export const nctBasket = process.env.REACT_APP_NCT_BASKET;
 // expLedger is a feature flag indicating the app is connected to a testnet with experimental features (Hambach).
 export const expLedger = process.env.REACT_APP_EXP_LEDGER;
 
+// v4Ledger is a feature flag indicating the app is connected to a testnet with Regen Ledger v4.0.
+// It can be removed once hambach/redwood/mainnet upgrades to v4.0.
+export const v4Ledger = process.env.REACT_APP_V4_LEDGER;
```
Comment (mhagel) [inline web-registry/src/lib/ledger.ts:11]: We will be able to delete these temporary V4 env vars. But not necessary right now.

File: web-registry/src/features/ecocredit/CreateBatchBySteps/CreateBatchMultiStepForm/CreditBasics.tsx [hunk 4]
```diff
@@ -76,88 +73,59 @@ export const validationSchemaFields = {
     .required(requiredMessage)
     .typeError(invalidDate)
     .test({ ...isPastDateTest }),
-  metadata: Yup.object().when('classId', {
-    is: 'C01',
+  metadata: Yup.object().when('projectId', {
+    // TODO: this is a hack to make V4 testnet work. C02 is a copy of C01 on V4. In PROD/mainnet, this should only be C01
+    is: (val: string) => val?.startsWith('C01') || val?.startsWith('C02'),
```
Comment (mhagel) [inline web-registry/src/features/ecocredit/CreateBatchBySteps/CreateBatchMultiStepForm/CreditBasics.tsx:78]:
```suggestion
    is: (val: string) => val?.startsWith('C01'),
```

File: web-registry/src/features/ecocredit/CreateBatchBySteps/CreateBatchMultiStepForm/CreditBasics.tsx [hunk 5]
```diff
@@ -76,88 +73,59 @@ export const validationSchemaFields = {
     .required(requiredMessage)
     .typeError(invalidDate)
     .test({ ...isPastDateTest }),
-  metadata: Yup.object().when('classId', {
-    is: 'C01',
+  metadata: Yup.object().when('projectId', {
+    // TODO: this is a hack to make V4 testnet work. C02 is a copy of C01 on V4. In PROD/mainnet, this should only be C01
```
Comment (mhagel) [inline web-registry/src/features/ecocredit/CreateBatchBySteps/CreateBatchMultiStepForm/CreditBasics.tsx:77]:
```suggestion
```

**Review by mhagel — APPROVED** [review]
a couple things to update for redwood

**Comment by mhagel** [comment]
We might as well PR `v4` into `redwood`, yeah?

**Comment by wgwz** [comment]
> We might as well PR `v4` into `redwood`, yeah?

@mhagel we could do it that way if it’s easier but i think there will be merge conflicts, which are what’s taken care of here. if we do it as a PR, we’ll probably want to make sure that the v4 branch does not get deleted after doing v4 into redwood as a PR. because i’m thinking we probably would want to merge v4 into master after that, right? suppose we did delete v4, it would be a little awkward to open a PR from redwood into master in order to get the v4 changes.

in any case, we should probably make the tweaks you’ve suggested in this PR in the v4 branch, and then do the v4-redwood and v4-master merges. 

WDYT?

**Comment by mhagel** [comment]
@wgwz Yes I see the merge conflicts now. 

>in any case, we should probably make the tweaks you’ve suggested in this PR in the v4 branch, and then do the v4-redwood and v4-master merges.
WDYT?

Sounds good.

**Comment by mhagel** [comment]
Some QA: 
- Create Batch works
- Basket Take and Put work
- Batch tables work
- Activity page works

Not working:
- Create Sell Order (keplr popup works, but TX fails silently - no error message.)
- Create Project (Location bug - not new - I created a [PR](https://github.com/regen-network/regen-web/pull/1126) to fix in `v4`)
- Lots of broken links, but **this is expected**. We have mismatches between our staging offchain DB and on-chain items that were created without our UI. (for example, credit class pages with no metadata - this may also be realted to why sell orders are failing)

@erikalogie If you QA this, note what I already discovered.

**Comment by erikalogie** [comment]
@mhagel do we need separate issues for fixing the failing sell orders and the broken links?

**Comment by wgwz** [comment]
Closing this for now, since as of earlier today we are planning to merge v4 into master first, and then proceed in the usual way by doing a backport of master into redwood.

**Pull request closed without merging**